lib/ocpp/v201/device_model.cpp

```diff
@@ -257,6 +293,43 @@ DeviceModel::get_report_data(const std::optional<ReportBaseEnum>& report_base,
     return report_data_vec;
 }
 
+std::vector<ReportData>
+DeviceModel::get_custom_report_data(const std::optional<std::vector<ComponentVariable>>& component_variables,
+                                    const std::optional<std::vector<ComponentCriterionEnum>>& component_criteria) {
+    std::vector<ReportData> report_data_vec;
+
+    for (auto const& [component, variable_map] : this->device_model) {
+        if (!component_criteria.has_value() or component_criteria_match_custom(component, component_criteria.value())) {
+
+            for (auto const& [variable, variable_meta_data] : variable_map) {
+                auto variable_ = variable;
+                auto component_ = component;
+
+                if (!component_variables.has_value() or
+                    (component_variables.has_value() and
```

Comment:
Don't need the second has_value since you already checked that and entered the loop if has_value is false.